webapp/src/main/java/com/box/l10n/mojito/pagerduty/PagerDutyClient.java

```diff
@@ -0,0 +1,119 @@
+    // Constructor for unit tests
+    this.integrationKey = integrationKey;
+    this.httpClient = httpClient;
+  }
+
+  /**
+   * Trigger incident using a deduplication key and send PagerDutyPayload with it. The client will
+   * attempt to send the event request MAX_RETRIES times if an internal error is received from the
+   * server. If the max number of retries is reached, a bad request is received or the payload
+   * cannot be serialized a PagerDutyException is thrown.
+   */
+  public void triggerIncident(String dedupKey, PagerDutyPayload payload) throws PagerDutyException {
+    sendPayload(dedupKey, payload, PagerDutyPostData.EventAction.TRIGGER);
+  }
+
+  /**
+   * Resolve incident using a deduplication key. The client will attempt to send the event request
+   * MAX_RETRIES times if an internal error is received from the server. If the max number of
+   * retries is reached, a bad request is received or the payload cannot be serialized a
+   * PagerDutyException is thrown.
+   */
+  public void resolveIncident(String dedupKey) throws PagerDutyException {
+    sendPayload(dedupKey, null, PagerDutyPostData.EventAction.RESOLVE);
+  }
+
+  private void sendPayload(
+      String dedupKey, PagerDutyPayload payload, PagerDutyPostData.EventAction eventAction)
+      throws PagerDutyException {
+
+    if (integrationKey == null || integrationKey.isEmpty())
+      throw new PagerDutyException("Integration key should not be null or empty.");
+
+    if (dedupKey == null || dedupKey.isEmpty())
+      throw new PagerDutyException("Deduplication key should not be null or empty.");
+
+    PagerDutyPostData postBody = new PagerDutyPostData(integrationKey, dedupKey);
+    postBody.setPayload(payload);
+    postBody.setEventAction(eventAction);
+
+    try {
+      HttpRequest request = buildRequest(postBody.serialize());
+      sendRequest(request);
+    } catch (JsonProcessingException e) {
+      throw new PagerDutyException("Failed to serialize PagerDutyPostRequest to a JSON string.", e);
+    }
+  }
+
+  private void sendRequest(HttpRequest request) throws PagerDutyException {
+    // Retry sending PD notifications if they fail
+    for (int attempt = 1; attempt <= MAX_RETRIES; attempt++) {
+      try {
+        HttpResponse<String> response =
+            httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+
+        int statusCode = response.statusCode();
+        if (statusCode == 200 || statusCode == 202) return;
+
+        // Either the max retries was hit or a bad request, there is no recovering from a 400 status
+        // code, exit without retries
+        if (attempt == MAX_RETRIES || statusCode == 400) {
+          logger.error(MessageFormat.format(exceptionTemplate, statusCode, response.body()));
+          throwFormattedException(statusCode, response.body());
+        }
+      } catch (IOException | InterruptedException e) {
+        if (attempt == MAX_RETRIES)
```

Comment:
I might be mistaken but does this logic make sense? 
The `attempt++` might not have run yet if the exception happened? (it happens after the loop body execution)

Reply:
This logic does make sense and there is a unit test to back it that forcefully throws IOException and counts how many times that the request was attempted to be sent. 

If a PagerDutyException is thrown the whole method will exit, but IOException / InterruptedException is thrown by the HttpClient and caught here, it will roll back into the for loop in which the attempt is incremented.

Reply:
Oh right. You start from 1. Yes okay. I missed that. This will work